### ISSUE TITLE
Normalization/ Normalized Metrics

### ISSUE THREAD

**Issue by JonathanSchmidt1** [Fung-Lab/MatterTune#15]
Hi, 

I had two questions. One is if you could please provide an example config. json for 
```
python -m mattertune.normalization \
    config.json \
    energy \
    references.json \
    --reference-model linear
```
the other one would be how to log normalized metrics, e.g. per_atom energy errors.
thank you so much.

**Comment by Lingyu-Kong** [comment]
Hi Jonathan, 

The MatterTune platform uses `MatterTunerConfig` as the central configuration object for every fine-tuning experiment. Within it you can specify the model, dataset, training hardware, fine-tuning algorithm, and more. A particularly convenient feature of the class is the `to_json_file()` method, which exports all settings to a JSON file—the `config.json` you asked about. Below we provide a brief showcase that demonstrates how to create and export such a file.

```
from __future__ import annotations

import logging
from pathlib import Path
import rich

import nshutils as nu
from lightning.pytorch.strategies import DDPStrategy

import mattertune.configs as MC
from mattertune import MatterTuner
from mattertune.configs import WandbLoggerConfig
from mattertune.backbones.jmp.model import get_jmp_s_lr_decay
from mattertune.backbones import (
    MatterSimM3GNetBackboneModule,
    JMPBackboneModule,
    ORBBackboneModule,
    EqV2BackboneModule,
)

logging.basicConfig(level=logging.ERROR)
nu.pretty()



def main():
    def hparams():
        hparams = MC.MatterTunerConfig.draft()
        hparams.model = MC.MatterSimBackboneConfig.draft()
        hparams.model.graph_convertor = MC.MatterSimGraphConvertorConfig.draft()
        hparams.model.pretrained_model = "MatterSim-v1.0.0-1M"
        hparams.model.optimizer = MC.AdamConfig(lr=8e-5)

        # Add model properties
        hparams.model.properties = []
        energy = MC.EnergyPropertyConfig(
            loss=MC.MSELossConfig(), loss_coefficient=1.0
        )
        hparams.model.properties.append(energy)
        forces = MC.ForcesPropertyConfig(
            loss=MC.MSELossConfig(), conservative=True, loss_coefficient=1.0
        )
        hparams.model.properties.append(forces)

        ## Data Hyperparameters
        hparams.data = MC.ManualSplitDataModuleConfig.draft()
        hparams.data.train = MC.XYZDatasetConfig.draft()
        hparams.data.train.src = "./data/train_water_1000_eVAng.xyz"
        hparams.data.train.down_sample = 30
        hparams.data.train.down_sample_refill = True
        hparams.data.validation = MC.XYZDatasetConfig.draft()
        hparams.data.validation.src = "./data/val_water_1000_eVAng.xyz"
        hparams.data.batch_size = 16

        # ## Trainer Hyperparameters
        # hparams.trainer = MC.TrainerConfig.draft()
        # hparams.trainer.max_epochs = 100
        # hparams.trainer.accelerator = "gpu"
        # hparams.trainer.devices = [0, 1, 2, 3]
        # hparams.trainer.strategy = "ddp"
        # hparams.trainer.gradient_clip_algorithm = "norm"
        # hparams.trainer.gradient_clip_val = 1.0
        # hparams.trainer.precision = "32"

        # # Additional trainer settings
        # hparams.trainer.additional_trainer_kwargs = {
        #     "inference_mode": False,
        # }

        hparams = hparams.finalize(strict=False)
        return hparams

    mt_config = hparams()
    
    ## NOTE: 
    ## Export the config to a json file before running the training
    ## Actually the trainer setting is not necessary so you can skip lines from 55 to 68
    mt_config.to_json_file(
        "./config.json",
    )
    
    # model, trainer = MatterTuner(mt_config).tune()
    

if __name__ == "__main__":
    main()
```

However, some highly complex experiment setups can’t be fully serialized, so the export may occasionally fail. If your only goal is to obtain a config.json that works with mattertune.normalization, simply follow the showcase and include just the model and data settings; exporting this minimalist JSON always succeeds.

We also offer an alternative way to obtain normalization information in `examples/water-thermodynamics/energy_reference.py`, hope that could help you too.

Finally, about recording normalized metrics: normalized values can end up in non-intuitive units, so before logging we denormalize both the predictions and the ground truths, ensuring that all reported metrics stay in standard units (e.g., eV / structure, eV / Å). The loss, however, is still computed on the normalized quantities, so the training process itself is unaffected.

**Comment by JonathanSchmidt1** [comment]
Thank you so much for the quick response! That perfectly solved the first question.
Concerning the metrics for energy errors if you have varying numbers of atoms a not normalized energy errors are quite difficult to work with. It would be great if you could point me in the right direction to add it.
I could add it by hand after the denormalization in _common step but I am sure there is a more elegant way.

**Comment by Lingyu-Kong** [comment]
And if you want to use multiple normalizer, you can just set them as a list, then the normalization will be conducted following its order in the list. For example:

```
## Add Normalization for Energy
        hparams.model.normalizers = {
            "energy": [
                MC.PerAtomReferencingNormalizerConfig("reference.json"),
                MC.PerAtomNormalizerConfig(),
            ]
        }
```

This example will firstly normalize energy based on ridge reference and then by numbers of atoms.
